### DISCUSSION TITLE
Dynamic price & cost calculation

### DISCUSSION THREAD

**Discussion by PlusA2M** [saleor/saleor#7718] [Q&A] [answered]
### What I want to achieve?
A common case for shop that selling custom made products, we’re trying to achieve: 
Calculating the cost and price dynamically depending on the product’s attribute/meta data rather than having all cost & price for all variant combinations which is too much almost impossible.

### What I call for help?
Does saleor has any webhooks that we could try develop an integration/app/plugin for that? Or which way should I go?

### Real world use-case
For example, we’re selling tailor-made product like name cards printing, we have our own formula for calculating the cost and price for any quantity and paper material combinations, in this case we don’t want to add product variants for all quantity and paper material, that’s just too much variants (that’s why we have our formula).

**Comment by karolkielecki** [comment, marked as answer]
@PlusA2M Hi there! I think it's not possible to create an app/plugin that introduces dynamic pricing at the moment. It requires bigger changes in the API. It's a cool feature--I'll discuss it with the team, but we have other priorities for now, e.g., Saleor 3.0 and other features you can find in our [roadmap](https://github.com/mirumee/saleor/projects/12).

**Reply by PlusA2M** [reply]
Thanks for your help! ❤️ So may I know is there any API for me to like modify/update the price of the order item on or before the customer entering the checkout process? I think we do want to use Saleor so much but our company are selling custom made products so that would be a must to us, and I was just thinking about if there are any other indirect way to do this at the moment?

For example, I could develop an function to calculate the pricing based on the meta-data of the product since we will develop our own frontend, and when the customer go to checkout or something related, we just make us of Saleor present's API to update the price in the cart/order.

**Reply by karolkielecki** [reply]
I think, unfortunately, it's a bigger feature that needs development and discussion of what has to be changed and added on our side. Also, it would require some changes on the dashboard side--and we do not recommend modifying the dashboard. Of course, you can always fork the code and modify it to your needs, but it won't be easy to stay up to date with the next Saleor releases.

**Reply by PlusA2M** [reply]
Gotcha, can't wait to hear the next update! I think we will be working on the 3.0 beta version and try to do some add-ons modifications just for this feature, hope that's go well. Any documentation reading suggestion would be appreciated!

**Reply by neuged** [reply]
Interesting! We thought about trying our hands on a plugin of that kind ourselves. From a cursory look at the data model we thought it might be possible to create a ProductVariant on the fly with certain numeric Attributes set to the user's wishes and a price for the variant based on a custom calculation. There might however be implications to this that we do not see.

**Reply by guyathomas** [reply]
@neuged did you end up exploring this?

**Reply by neuged** [reply]
Yes, we did explore this a bit. As a proof of concept we wrote a small Saleor App. The app exposes a public route where the frontend client can send product dimensions to get a price calculation. Optionally the client can ask the app to create a product variant with the given dimensions. The app has a token from our saleor instance and is configured with the MANAGE_PRODUCTS permission, so it can create the product variant using a mutation. It then sends back the id of the new variant, which the client can then use further. It's not production ready yet, but we plan to use it in the future.

**Reply by PlusA2M** [reply]
> @PlusA2M Hi there! I think it's not possible to create an app/plugin that introduces dynamic pricing at the moment. It requires bigger changes in the API. It's a cool feature--I'll discuss it with the team, but we have other priorities for now, e.g., Saleor 3.0 and other features you can find in our [roadmap](https://github.com/mirumee/saleor/projects/12). 

@karolkielecki We've seen 3.1 is released! Great applause and cheers! As we've seen that some of other [discussions](https://github.com/saleor/saleor/discussions/8376) here talking with similar feature, we're designing an plugin to achieve the desired feature, however, calculating price dynamically would affect parts not only to checkout, but also product variants, we're  wondering if there are an event/webhook fired exactly on every times the price is calculated? Thanks for any suggestions in advance!

Issues/discussions related to dynamic product & shipping pricing:
- https://github.com/saleor/saleor/issues/5115
- https://github.com/saleor/saleor/issues/8619
- https://github.com/saleor/saleor/discussions/8376
- https://github.com/saleor/saleor/issues/2029
- https://github.com/saleor/saleor/discussions/8336
- https://github.com/saleor/saleor/issues/1080
- https://github.com/saleor/saleor/issues/1093

**Comment by timuric** [comment]
Dynamic price calculation is available in 3.4
https://docs.saleor.io/docs/3.x/developer/checkout/lines#setting-custom-prices-for-checkout-lines

**Reply by PlusA2M** [reply]
Thanks @timuric ! We will look into it soon.

**Reply by ivanakcheurov** [reply]
@timuric , the page doesn’t have neither the section nor the anchor related to setting custom prices anymore.
I could find no synchronous webhook for this either.
Was this functionality removed or just moved somewhere else?

**Reply by timuric** [reply]
Indeed in the meantime the docs structure changed, here is the link 
https://docs.saleor.io/docs/3.x/developer/checkout/lines#setting-custom-prices-for-checkout-lines

**Comment by yagiz-aydin** [comment]
Hello @timuric, ✋ 

This is a case study for now, I'm checking all the best practices for the case and I like the Saelor's everything ❤️ For my shop, I want that selling custom products with some managable priced different attributes. For example: I wanna sell a tablet with changable ram, screen size... so user can change this dynamic attributes on the Product detail screen after all customizable section are done User can add this product to the basket. Is this feature is already added or not capable for now ?

**Reply by timuric** [reply]
there are many saleor based stores that do complex configurators, you would have to use custom logic for price calculation and overwrite the checkout line using https://docs.saleor.io/docs/3.x/developer/checkout/lines#setting-custom-prices-for-checkout-lines
